### PR TITLE
feat: add format validation to text and textarea

### LINKED ISSUE #122: Add support for format validations
https://github.com/projectcaluma/caluma/issues/360 implements format validations. This issue can be used to track the frontend integration:

- [ ] Form builder integration
- [ ] Form rendering integration

### PR DESCRIPTION
**Pull request by JohnL17**
Resolves #122

### CODE REVIEW

**Comment by anehx** [comment]
@JohnL17 I'll have a look at this when #376 landed in master since it's likely that there will be changes necessary..

File: addon/components/cfb-form-editor/question/validation.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  classNames: ["format-validators"],
+  layout,
+
+  apollo: service(),
+
+  async didReceiveAttrs() {
+    this._super(...arguments);
+
+    await this.get("availableFormatValidators").perform();
+  },
```
Comment (anehx) [inline addon/components/cfb-form-editor/question/validation.js]: 1. I think the `didReceiveAttrs` is the wrong hook for this, since it will perform every time an argument changes but the task has nothing to do with those. I'd prefer using `init`.
2. You don't need to await the task. This will only block rendering which is not very good UX

File: addon/components/cfb-form-editor/question/validation.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  availableFormatValidators: task(function*() {
+    const formatValidators = yield this.get("apollo").watchQuery(
+      { query: allFormatValidatorsQuery, fetchPolicy: "cache-and-network" },
+      "allFormatValidators.edges"
+    );
+    return formatValidators.map(edge => edge.node);
+  }).restartable(),
```
Comment (anehx) [inline addon/components/cfb-form-editor/question/validation.js]: `.drop()` might be better, since you don't really want this to run multiple times

File: addon/components/cfb-form-editor/question/validation.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  validators: computed(
+    "availableFormatValidators.lastSuccessful.value",
+    function() {
+      if (this.availableFormatValidators.lastSuccessful) {
+        return this.availableFormatValidators.lastSuccessful.value;
+      } else {
+        return [];
+      }
+    }
+  ),
```
Comment (anehx) [inline addon/components/cfb-form-editor/question/validation.js]: `if ... return ... else` is an anti-pattern

```js
if (!this.availableFormatValidators.lastSuccessful) {
  return [];
}

return this.availableFormatValidators.lastSuccessful.value;
```

Comment (anehx) [reply addon/components/cfb-form-editor/question/validation.js]: Or use `getWithDefault` like below

File: addon/components/cfb-form-editor/question/validation.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+  selected: computed("value", "validators", function() {
+    return this.getWithDefault("validators", []).filter(validator =>
```
Comment (anehx) [inline addon/components/cfb-form-editor/question/validation.js]: `getWithDefault` is not needed, since you catch that case in `validators`

File: addon/lib/field.js [hunk 5]
```diff
@@ -72,21 +73,50 @@ export default Base.extend(Evented, {
   saveDocumentDateAnswerMutation,
   saveDocumentTableAnswerMutation,
 
+  /**
+   * The Apollo GraphQL service for making requests
+   *
+   * @property {ApolloService} apollo
+   * @accessor
+   */
```
Comment (anehx) [inline addon/lib/field.js]: In the lib we only want the public API or complex private methods/properties documented, this is kind of obvious what it is.

File: addon/lib/field.js [hunk 6]
```diff
@@ -72,21 +73,50 @@ export default Base.extend(Evented, {
...
   apollo: service(),
+
+  /**
+   * The translation service
+   *
+   * @property {IntlService} intl
+   * @accessor
+   */
```
Comment (anehx) [inline addon/lib/field.js]: Same here

File: addon/lib/field.js [hunk 7]
```diff
@@ -72,21 +73,50 @@ export default Base.extend(Evented, {
...
   intl: service(),
   calumaStore: service(),
 
+  /**
+   * The Validator service
+   *
+   * @property {ValidatorService} validator
+   * @accessor
+   */
```
Comment (anehx) [inline addon/lib/field.js]: Same here

File: addon/lib/field.js [hunk 8]
```diff
@@ -72,21 +73,50 @@ export default Base.extend(Evented, {
...
+  validator: service(),
+
+  /**
+   * Initialize function which validates the passed arguments and sets an
+   * initial state of errors.
+   *
+   * @method init
+   * @internal
+   */
```
Comment (anehx) [inline addon/lib/field.js]: Same here

File: addon/lib/field.js [hunk 9]
```diff
@@ -72,21 +73,50 @@ export default Base.extend(Evented, {
...
   init() {
     this._super(...arguments);
 
+    this.validator.getValidators();
```
Comment (anehx) [inline addon/lib/field.js]: This doesn't do anything :wink:

File: addon/lib/field.js [hunk 10]
```diff
@@ -72,21 +73,50 @@ export default Base.extend(Evented, {
...
+
     this.setProperties({
       _errors: []
     });
   },
 
   /**
-   * The unique identifier for the field which consists of the documents pk and
-   * the questions pk separated by a colon.
+   * The ID of the field. Consists of the document ID and the question slug.
```
Comment (anehx) [inline addon/lib/field.js]: Why'd you change that?

File: addon/lib/field.js [hunk 11]
```diff
@@ -141,6 +171,13 @@ export default Base.extend(Evented, {
     return answer.id ? this.calumaStore.push(answer) : answer;
   }),
 
+  registerDependentField(field, key) {
+    this.set(`dependentFields.${key}`, [
+      ...new Set(this.get(`dependentFields.${key}`)),
+      field
+    ]);
+  },
```
Comment (anehx) [inline addon/lib/field.js]: Obsolete, was removed in #376

File: addon/lib/field.js [hunk 12]
```diff
@@ -405,9 +442,15 @@ export default Base.extend(Evented, {
...
-    });
+    return [
+      ...this.validator.validate(
+        this.get("answer.value"),
+        this.getWithDefault("question.meta.formatValidators", [])
+      ),
+      validate("length", this.get("answer.value"), {
+        max: this.get("question.textMaxLength") || Number.POSITIVE_INFINITY
+      })
+    ];
```
Comment (anehx) [inline addon/lib/field.js]: You need to return a `Promise` here, the `validate` task does not support arrays as return value of a validation function

File: addon/lib/field.js [hunk 13]
```diff
@@ -419,9 +462,15 @@ export default Base.extend(Evented, {
...
-    });
+    return [
+      ...this.validator.validate(
+        this.get("answer.value"),
+        this.getWithDefault("question.meta.formatValidators", [])
+      ),
+      validate("length", this.get("answer.value"), {
+        max: this.get("question.textareaMaxLength") || Number.POSITIVE_INFINITY
+      })
+    ];
```
Comment (anehx) [inline addon/lib/field.js]: Same here

File: addon/mirage-graphql/schema.graphql [hunk 14]
```diff
@@ -626,6 +626,7 @@ type Document implements Node {
     orderBy: [AnswerOrdering]
     questions: [ID]
   ): AnswerConnection
+  parentAnswers: [FormAnswer]
```
Comment (anehx) [inline addon/mirage-graphql/schema.graphql]: This doesn't exist anymore, removed in #376

File: addon/mirage-graphql/schema.graphql [hunk 15]
```diff
@@ -1050,6 +1051,22 @@ type Form implements Node {
...
+
+  """
+  The ID of the object.
+  """
+  id: ID!
+  question: Question!
+  value: Document!
+  meta: GenericScalar!
+  document: Document!
+}
```
Comment (anehx) [inline addon/mirage-graphql/schema.graphql]: Same here

File: addon/mirage-graphql/schema.graphql [hunk 16]
```diff
@@ -2138,6 +2176,23 @@ type SaveDocumentFloatAnswerPayload {
...
+
+type SaveDocumentFormAnswerPayload {
+  answer: Answer
+  clientMutationId: String
+}
+
 input SaveDocumentInput {
   id: String
   form: ID!
```
Comment (anehx) [inline addon/mirage-graphql/schema.graphql]: Did you run `yarn update-schema` with the **latest** container? This seems like a deprecated version...

File: addon/services/validator.js [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  },
+
+  /**
+   * Tests a value against one or multiple regular expressions.
+   *
+   * ```js
+   * this.validator.validate("foo@example.com", ["email", "lowercase"]);
+   * ```
+   * @param {String} value The value to be tested.
+   * @param {Array<String>} slugs A list of tests (via slug) to run.
```
Comment (anehx) [inline addon/services/validator.js]:
```suggestion
   * @param {String[]} slugs A list of tests (via slug) to run.
```

File: addon/services/validator.js [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+  _translations: computed("_fetchValidators.lastSuccessful.value", function() {
+    const translations = {};
+    this.getWithDefault("_fetchValidators.lastSuccessful.value", []).forEach(
+      edge => {
+        translations[edge.node.slug] = edge.node.errorMsg;
+      }
+    );
+    return translations;
+  })
+});
```
Comment (anehx) [inline addon/services/validator.js]: That service seems kind of strange to me, we'll discuss this in person

File: package.json [hunk 19]
```diff
@@ -49,22 +49,27 @@
     "ember-fetch": "^6.5.0",
     "ember-intl": "^4.0.0",
     "ember-math-helpers": "^2.10.0",
+    "ember-promise-helpers": "^1.0.6",
```
Comment (anehx) [inline package.json]: Where do you need this?

File: package.json [hunk 20]
```diff
@@ -49,22 +49,27 @@
...
+    "liquid-fire": "^0.29.5",
     "sass": "^1.18.0",
     "uuid": "^3.3.2",
     "slugify": "^1.3.4"
   },
   "devDependencies": {
     "@adfinis-sygroup/semantic-release-config": "2.1.3",
     "@babel/plugin-proposal-object-rest-spread": "7.4.4",
+    "@commitlint/cli": "7.6.1",
+    "@commitlint/config-conventional": "7.6.0",
```
Comment (anehx) [inline package.json]: Obsolete, `@adfinis-sygroup/semantic-release-config` lists those as dependency

File: package.json [hunk 21]
```diff
@@ -78,6 +83,7 @@
     "ember-cli-deploy-build": "1.1.1",
     "ember-cli-deploy-git": "1.3.3",
     "ember-cli-deploy-git-ci": "1.0.1",
+    "ember-cli-eslint": "5.1.0",
```
Comment (anehx) [inline package.json]: Obsolete, we use `eslint` directly

File: package.json [hunk 22]
```diff
@@ -102,7 +107,7 @@
     "eslint-config-prettier": "6.0.0",
     "eslint-plugin-ember": "6.7.0",
     "eslint-plugin-node": "9.1.0",
-    "eslint-plugin-prettier": "3.1.0",
+    "eslint-plugin-prettier": "2.6.0",
```
Comment (anehx) [inline package.json]: Why?

File: package.json [hunk 23]
```diff
@@ -112,6 +117,9 @@
     "typeface-oxygen": "0.0.72",
     "typeface-oxygen-mono": "0.0.72"
   },
+  "resolutions": {
+    "**/tough-cookie": "~2.4.0"
+  },
```
Comment (anehx) [inline package.json]: Why?

File: tests/dummy/app/routes/index.js [hunk 24]
```diff
@@ -40,7 +40,7 @@ export default Route.extend(RouteQueryManager, {
       {
         query: gql`
           query {
-            allDocuments(form: "main", orderBy: CREATED_AT_DESC) {
+            allDocuments(form: "test", orderBy: CREATED_AT_DESC) {
```
Comment (anehx) [inline tests/dummy/app/routes/index.js]: That seems weird

File: tests/integration/components/cf-field-test.js [hunk 25]
```diff
@@ -3,9 +3,11 @@ import { setupRenderingTest } from "ember-qunit";
 import { render, fillIn } from "@ember/test-helpers";
 import hbs from "htmlbars-inline-precompile";
 import Document from "ember-caluma/lib/document";
+import { setupMirage } from "ember-cli-mirage/test-support";
 
 module("Integration | Component | cf-field", function(hooks) {
   setupRenderingTest(hooks);
+  setupMirage(hooks);
```
Comment (anehx) [inline tests/integration/components/cf-field-test.js]: Why is this needed?

File: tests/integration/components/cf-field/input/date-test.js [hunk 26]
```diff
@@ -4,9 +4,11 @@ import { click, render } from "@ember/test-helpers";
 import moment from "moment";
 import hbs from "htmlbars-inline-precompile";
 import { Interactor } from "ember-pikaday/test-support";
+import { setupMirage } from "ember-cli-mirage/test-support";
 
 module("Integration | Component | cf-field/input/date", function(hooks) {
   setupRenderingTest(hooks);
+  setupMirage(hooks);
```
Comment (anehx) [inline tests/integration/components/cf-field/input/date-test.js]: Same here

File: tests/integration/components/cf-field/label-test.js [hunk 27]
```diff
@@ -3,9 +3,11 @@ import { setupRenderingTest } from "ember-qunit";
 import { render, settled } from "@ember/test-helpers";
 import hbs from "htmlbars-inline-precompile";
 import Document from "ember-caluma/lib/document";
+import { setupMirage } from "ember-cli-mirage/test-support";
 
 module("Integration | Component | cf-field/label", function(hooks) {
   setupRenderingTest(hooks);
+  setupMirage(hooks);
```
Comment (anehx) [inline tests/integration/components/cf-field/label-test.js]: Same here

File: tests/unit/lib/document-test.js [hunk 28]
```diff
@@ -2,9 +2,11 @@ import { module, test, skip } from "qunit";
 import { setupTest } from "ember-qunit";
 import Document from "ember-caluma/lib/document";
 import { settled } from "@ember/test-helpers";
+import { setupMirage } from "ember-cli-mirage/test-support";
 
 module("Unit | Library | document", function(hooks) {
   setupTest(hooks);
+  setupMirage(hooks);
```
Comment (anehx) [inline tests/unit/lib/document-test.js]: Same here

File: tests/unit/lib/field-test.js [hunk 29]
```diff
@@ -2,11 +2,13 @@ import { module, test } from "qunit";
 import { setupTest } from "ember-qunit";
 import { settled } from "@ember/test-helpers";
 import Field from "ember-caluma/lib/field";
+import { setupMirage } from "ember-cli-mirage/test-support";
 import Document from "ember-caluma/lib/document";
 import faker from "faker";
 
 module("Unit | Library | field", function(hooks) {
   setupTest(hooks);
+  setupMirage(hooks);
```
Comment (anehx) [inline tests/unit/lib/field-test.js]: Same here

File: tests/unit/lib/question-test.js [hunk 30]
```diff
@@ -1,9 +1,11 @@
 import { module, test } from "qunit";
 import { setupTest } from "ember-qunit";
 import Question from "ember-caluma/lib/question";
+import { setupMirage } from "ember-cli-mirage/test-support";
 
 module("Unit | Library | question", function(hooks) {
   setupTest(hooks);
+  setupMirage(hooks);
```
Comment (anehx) [inline tests/unit/lib/question-test.js]: Same here

File: translations/de-de.yaml [hunk 31]
```diff
@@ -38,6 +38,7 @@ caluma:
       notAnInteger: "Bitte geben Sie eine ganze Zahl ein"
       inclusion: "'{value}' ist kein gültiger Wert für dieses Feld"
       uploadFailed: "Beim Hochladen ist ein Fehler aufgetreten."
+      format: "{errorMsg}"
```
Comment (anehx) [inline translations/de-de.yaml]: That does not make sense, this will just render the passed argument

File: translations/en-us.yaml [hunk 32]
```diff
@@ -38,6 +38,7 @@ caluma:
       notAnInteger: "The value of this field must be an integer"
       inclusion: "'{value}' is not a valid value for this field"
       uploadFailed: "An error occured during upload."
+      format: "{errorMsg}"
```
Comment (anehx) [inline translations/en-us.yaml]: Same here

File: addon/components/cfb-form-editor/question.js [hunk 33]
```diff
@@ -292,6 +291,13 @@ export default Component.extend(ComponentQueryManager, {
 
       yield this.saveOptions.perform(changeset);
 
+      changeset.set(
+        "meta.formatValidators",
+        (changeset.get("meta.formatValidators") || []).map(
+          validator => validator.slug
+        )
+      );
```
[the remark this answers is not in the document]
Comment (anehx) [reply addon/components/cfb-form-editor/question.js]: Discussed: remove this

File: addon/components/cfb-form-editor/question/validation.js [hunk 34]
```diff
@@ -0,0 +1,51 @@
...
+    function() {
+      if (this.availableFormatValidators.lastSuccessful) {
+        return this.availableFormatValidators.lastSuccessful.value;
+      } else {
+        return [];
+      }
+    }
+  ),
+
+  selected: computed("value", "validators", function() {
```
Comment (anehx) [inline addon/components/cfb-form-editor/question/validation.js]:
```suggestion
  selected: computed("value.[]", "validators.@each.slug", function() {
```

File: addon/lib/field.js (hunk 9 above)
Comment (anehx) [reply addon/lib/field.js]: Discussed: remove this

File: addon/lib/field.js (hunk 10 above)
Comment (anehx) [reply addon/lib/field.js]: Discussed: remove this

File: addon/lib/field.js (hunk 11 above)
Comment (anehx) [reply addon/lib/field.js]: Discussed: remove this

File: addon/lib/field.js (hunk 12 above)
Comment (anehx) [reply addon/lib/field.js]: Discussed: return a single value here

File: addon/services/validator.js [hunk 35]
```diff
@@ -0,0 +1,80 @@
...
+   * ```js
+   * this.validator.validate("foo@example.com", ["email", "lowercase"]);
+   * ```
+   * @param {String} value The value to be tested.
+   * @param {Array<String>} slugs A list of tests (via slug) to run.
+   */
+  validate(value, slugs) {
+    if (!Array.isArray(slugs)) {
+      slugs = [slugs];
+    }
```
Comment (anehx) [inline addon/services/validator.js]: Discussed: drop this

File: addon/services/validator.js [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+      assert(`No validator found with the slug "${slug}".`, validator);
+
+      return validator.test(value)
+        ? true
+        : {
+            type: "format",
+            message: undefined,
+            context: { errorMsg: this.getText(slug) },
+            value
+          };
```
Comment (anehx) [inline addon/services/validator.js]:
```js
return validator.test(value) || // ...
```

File: addon/services/validator.js [hunk 37]
```diff
@@ -0,0 +1,80 @@
...
+    });
+  },
+
+  getValidators() {
+    return this.get("_validators");
+  },
```
Comment (anehx) [inline addon/services/validator.js]: Discussed: remove this

File: package.json [hunk 38]
```diff
@@ -49,22 +49,27 @@
...
     "ember-uikit": "^0.8.1",
     "ember-validated-form": "^2.0.0-alpha.5",
     "faker": "4.1.0",
+    "ember-power-select": "2.3.4",
     "graphql": "^14.2.1",
     "graphql-iso-date": "^3.6.1",
     "graphql-tag": "^2.10.1",
     "graphql-tools": "^4.0.4",
     "jexl": "TomFrost/Jexl#2866a4aca9c46114ecc0d342942a7732763fd151",
+    "liquid-fire": "^0.29.5",
```
Comment (anehx) [inline package.json]: Drop this

File: tests/integration/components/cf-field-test.js (hunk 25 above)
Comment (anehx) [reply tests/integration/components/cf-field-test.js]: https://guides.emberjs.com/v3.10.0/testing/testing-components/#toc_stubbing-services

File: tests/unit/services/validator-test.js [hunk 39]
```diff
@@ -0,0 +1,28 @@
...
+    let service = this.owner.lookup("service:validator");
+    assert.ok(service);
+  });
+
+  test("slugs are in a string not in an array", async function(assert) {
+    assert.expect(1);
+    this.server.logging = true;
+    let service = this.owner.lookup("service:validator");
+    await settled();
+    let result = service.validate("test@test.com", "email");
```
Comment (anehx) [inline tests/unit/services/validator-test.js]: The second argument has to be an array

File: tests/integration/components/cf-field-test.js (hunk 25 above)
Comment (anehx) [reply tests/integration/components/cf-field-test.js]: Locate the stub in `/tests/helpers/validator-service-stub.js`

File: addon/lib/field.js (hunk 12 above)
Comment (JohnL17) [reply addon/lib/field.js]: Florian said arrays are fine.

We already have functions returning an array ([L527](https://github.com/projectcaluma/ember-caluma/blob/master/addon/lib/field.js#L527)) and the return value from the function is transformed to an array anyway ([L376](https://github.com/projectcaluma/ember-caluma/blob/master/addon/lib/field.js#L376)).

File: tests/integration/components/cfb-form-editor/question/validation-test.js [hunk 40]
```diff
@@ -0,0 +1,69 @@
...
+      let service = this.owner.lookup("service:validator");
+      await settled();
+      let error = service.validators.find(i => i.slug === "email").errorMsg;
+
+      await render(hbs`{{cf-field field=field}}`);
+
+      await fillIn("input", "Test");
+
+      assert.dom("span.validation-errors").hasText(error);
+    });
```
Comment (czosel) [inline tests/integration/components/cfb-form-editor/question/validation-test.js]: Shall we also add a test for the successful case (valid email address)?

File: translations/de-de.yaml [hunk 41]
```diff
@@ -87,6 +88,9 @@ caluma:
       supportsMarkdownPrefix: "Dieses Feld unterstützt"
       markdown: "Markdown"
       dataSource: "Datenquelle"
+      formatValidators: "Validierung"
+      chooseValidation: "Wähle eine Validierung"
```
Comment (czosel) [inline translations/de-de.yaml]: I'd suggest either the formal "Bitte wählen Sie eine Validierung" or porbably even better the already existing "-- bitte wählen --" (`choose`, L103).

File: translations/en-us.yaml [hunk 42]
```diff
@@ -87,6 +88,8 @@ caluma:
       supportsMarkdownPrefix: "This field supports"
       markdown: "Markdown"
       dataSource: "Data source"
+      formatValidators: "Validation"
+      chooseValidation: "Choose a validation"
```
Comment (czosel) [inline translations/en-us.yaml]: Same here

File: tests/integration/components/cfb-form-editor/question/validation-test.js (hunk 40 above)
Comment (JohnL17) [reply tests/integration/components/cfb-form-editor/question/validation-test.js]: Done

File: translations/de-de.yaml (hunk 41 above)
Comment (JohnL17) [reply translations/de-de.yaml]: Done

File: translations/en-us.yaml (hunk 42 above)
Comment (JohnL17) [reply translations/en-us.yaml]: Done

File: tests/integration/components/cfb-form-editor/question/validation-test.js [hunk 43]
```diff
@@ -65,5 +65,18 @@ module(
...
+      assert.expect(1);
+      let service = this.owner.lookup("service:validator");
+      await settled();
+      let error = service.validators.find(i => i.slug === "email").errorMsg;
+
+      await render(hbs`{{cf-field field=field}}`);
+
+      await fillIn("input", "test@test.com");
+
+      assert.dom("span.validation-errors").hasText(error);
```
Comment (czosel) [inline tests/integration/components/cfb-form-editor/question/validation-test.js]: Aren't you asserting that an error is displayed here?

**Comment by czosel** [comment]
LGTM now. Can you fix the conflicts?

File: tests/integration/components/cf-field-test.js [hunk 44]
```diff
@@ -30,9 +30,15 @@ module("Integration | Component | cf-field", function(hooks) {
...
-            formatValidators: ["email"]
-          },
+          meta: { formatValidators: ["email"] },
+          __typename: "TextQuestion"
+        },
+        {
+          slug: "question-3",
+          label: "Test3",
+          isRequired: "true",
+          isHidden: "true",
```
Comment (czosel) [inline tests/integration/components/cf-field-test.js]: Nitpick: Shouldn't this be `false`? (same above)

Comment (JohnL17) [reply tests/integration/components/cf-field-test.js]: Yes you are right. I didn't recognize it.
